cmd/kubeadm/app/phases/controlplane/manifests.go

```diff
@@ -96,6 +97,15 @@ func CreateStaticPodFiles(manifestDir, patchesDir string, cfg *kubeadmapi.Cluste
 	klog.V(1).Infoln("[control-plane] getting StaticPodSpecs")
 	specs := GetStaticPodSpecs(cfg, endpoint)
 
+	var usersAndGroups *users.UsersAndGroups
+	var err error
+	if features.Enabled(cfg.FeatureGates, features.RootlessControlPlane) && !isDryRun {
```

Comment:
just to double check something.

if this github search is correct, CreateStaticPodFiles is called in the right places:
https://github.com/kubernetes/kubernetes/search?q=CreateStaticPodFiles
but it is also called from CreateInitStaticPodManifestFiles.

ideally we would like to create users and groups once per: init,join,upgrade operation.
is that true based on your observations?

Reply:
Unfortunately no! 
That is because we call CreateStaticPodFiles per component:-
see https://github.com/kubernetes/kubernetes/blob/f3371ced8593498a472452805b8ba212d67ea47d/cmd/kubeadm/app/cmd/phases/init/controlplane.go#L149
and
https://github.com/kubernetes/kubernetes/blob/f3371ced8593498a472452805b8ba212d67ea47d/cmd/kubeadm/app/cmd/phases/join/controlplaneprepare.go#L190
Only in upgrade which is when we call CreateInitStaticPodManifestFiles we pass all components in a single call to CreateStaticPodFiles.

I think we should change this behavior in the callers of CreateStaticPodFiles where possible in the future.

Reply:
Instead of syn.Mutex I think we should use syn.Once for scenarios like this , wdyt?

Reply:
sync.Once works too, if done properly.

Reply:
> sync.Once works too, if done properly.

but moving to `kubeadm/app/util/staticpod/utils.go` still makes sense.

Reply:
Done.